### PR TITLE
Yf geometrical downsampling

### PR DESCRIPTION
**Pull request by yfarjoun**
- Moved PgIdGenerator from Picard-public to here 
- Fixed a small bug in SAMRecordSetBuilder (didn't add the NM tag which then fails validation)

Needed for a new tool in Picard (Position-based downsampling)

### CODE REVIEW

File: src/java/htsjdk/samtools/SAMRecordSetBuilder.java [hunk 1]
```diff
@@ -346,6 +347,7 @@ public void addPair(final String name, final int contig, final int start1, final
         end2.setAlignmentStart(start2);
         end2.setReadNegativeStrandFlag(true);
         end2.setCigarString(readLength + "M");
+        end2.setAttribute(ReservedTagConstants.NM,0);
```
Comment (nh13) [inline src/java/htsjdk/samtools/SAMRecordSetBuilder.java]: I don't think this is the only place where ```NM``` is not set (see the other ```addPair``` methods).  Do you want to add it to ```createReadNoFlag```?

Comment (yfarjoun) [inline src/java/htsjdk/samtools/SAMRecordSetBuilder.java]: impossible. :-/

Comment (yfarjoun) [inline src/java/htsjdk/samtools/SAMRecordSetBuilder.java]: OK. fixed it. look again?

**Comment by yfarjoun** [comment]
Got verbal :+1: from @nh13 

will merge once tests pass.

**Pull request merged**